### PR TITLE
Add the 3d avg pool for video related model

### PR DESCRIPTION
**Pull request by lly-zero-one**
```
import torch, time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    print('****', str(dtype), '*****')
    x = torch.rand(1, 5, 56, 56, 256)

    q_x = torch.quantize_per_tensor(x, 0.5, 1, dtype)
    q_x = q_x.permute([0, 4, 1, 2, 3])

    x = x.permute([0, 4, 1, 2, 3])

    NITER = 10

    s = time.time()
    for i in range(NITER):
        float_out = torch.nn.functional.avg_pool3d(x, kernel_size=3, stride=None, padding=0)
    time_per_iter_float = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.nn.quantized.functional.avg_pool3d(q_x, kernel_size=3, stride=None, padding=0)
    time_per_iter_quant = (time.time() - s) / NITER
    print('time/iter ms (float)', 'time/iter ms (quant)', 'quant/float', sep='\t')
    print(time_per_iter_float * 1000, time_per_iter_quant * 1000, time_per_iter_quant / time_per_iter_float, sep='\t')
```

```
**** torch.qint8 *****
time/iter ms (float)  time/iter ms (quant)  quant/float
16.286182403564453  0.7308721542358398  0.04487682479080417
**** torch.quint8 *****
time/iter ms (float)  time/iter ms (quant)  quant/float
15.364313125610352  0.6497383117675781  0.042288796541418254
**** torch.qint32 *****
time/iter ms (float)  time/iter ms (quant)  quant/float
15.649032592773438  13.879132270812988  0.8869003363966556
```

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -782,6 +793,121 @@ void qavg_pool2d_nhwc_kernel(
...
+          int64_t dstart = od * dD - padD;
+          int64_t hstart = oh * dH - padH;
+          int64_t wstart = ow * dW - padW;
+
+          int64_t dend = std::min(dstart + kD, inputDepth + padD);
+          int64_t hend = std::min(hstart + kH, inputHeight + padH);
+          int64_t wend = std::min(wstart + kW, inputWidth + padW);
+          int64_t pool_size = (dend - dstart) * (hend - hstart) * (wend - wstart);
+
+          dstart = std::max(dstart, (int64_t)0);
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: `(int64_t)0` -> `static_cast<int64_t>(0)` or `0L`?

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 2]
```diff
@@ -782,6 +793,121 @@ void qavg_pool2d_nhwc_kernel(
...
+            } else {
+              divide_factor = size;
+            }
+          }
+
+          int64_t c = 0;
+          // For int8 quantization, we implicitly use int32 as accumulation
+          // Or else, it will go to the slow path
+          // TODO: support 16bit, 32bit, and etc.
+          float multiplier = qx.q_scale() / qy.q_scale() / divide_factor;
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Do we need to add a assert to make sure `divide_factor` is not 0?

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 3]
```diff
@@ -40,15 +40,15 @@ template <typename scalar_t>
 static void adaptive_avg_pool2d_single_out_frame(
     scalar_t* input_p,
     scalar_t* output_p,
-    int64_t sizeD,
+    int64_t sizeC,
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: Wonder the reason for changing the naming? Is it a convention? `C` means channel, and `D` is instead used for depth in the 3D pooling?

File: aten/src/ATen/native/quantized/cpu/q_avgpool3d.cpp [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+            }
+
+            int64_t kx, ky, kz;
+            for (kz = dstart; kz < dend; kz++) {
+              for (ky = hstart; ky < hend; ky++) {
+                for (kx = wstart; kx < wend; kx++)
+                  sum_int += (ptr_input + kz * inputHeight * inputWidth + ky * inputWidth + kx)->val_;
+              }
+            }
+            float multiplier = input.q_scale() / output.q_scale() / divide_factor;
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/q_avgpool3d.cpp]: Ditto: assert divisor is not 0.

File: aten/src/ATen/native/quantized/cpu/q_avgpool3d.cpp [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+          for (xx = 0; xx < outputWidth; xx++) {
+            /* Compute the mean of the input image... */
+            int64_t dstart = zz * dD - padD;
+            int64_t hstart = yy * dH - padH;
+            int64_t wstart = xx * dW - padW;
+            int64_t dend = std::min(dstart + kD, inputDepth + padD);
+            int64_t hend = std::min(hstart + kH, inputHeight + padH);
+            int64_t wend = std::min(wstart + kW, inputWidth + padW);
+            int64_t pool_size = (dend - dstart) * (hend - hstart) * (wend - wstart);
+            dstart = std::max(dstart, (int64_t)0);
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/q_avgpool3d.cpp]: Ditto: C++ style static_cast

File: aten/src/ATen/native/quantized/cpu/q_avgpool3d.cpp [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+          b * nInputPlane * outputWidth * outputHeight * outputDepth +
+          k * outputWidth * outputHeight * outputDepth;
+      const scalar_t* ptr_input = input_data +
+          b * nInputPlane * inputWidth * inputHeight * inputDepth +
+          k * inputWidth * inputHeight * inputDepth;
+      auto minimum =
+          std::numeric_limits<typename scalar_t::underlying>::lowest();
+      auto maximum = std::numeric_limits<typename scalar_t::underlying>::max();
+
+      for (zz = 0; zz < outputDepth; zz++) {
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/q_avgpool3d.cpp]: replace zz, yy, xx with variable names like od, oh, ow?

**Comment by jianyuh** [comment]
Please check the test failure:
```
Feb 14 11:22:37 ======================================================================
Feb 14 11:22:37 ERROR [3.030s]: test_avg_pool3d_nhwc (__main__.TestQuantizedOps)
Feb 14 11:22:37 ----------------------------------------------------------------------
Feb 14 11:22:37 Traceback (most recent call last):
Feb 14 11:22:37   File "test_quantized.py", line 780, in test_avg_pool3d_nhwc
Feb 14 11:22:37     min_side=5, max_side=10),
Feb 14 11:22:37   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 1116, in wrapped_test
Feb 14 11:22:37     raise the_error_hypothesis_found
Feb 14 11:22:37   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 1071, in wrapped_test
Feb 14 11:22:37     state.run_engine()
Feb 14 11:22:37   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 789, in run_engine
Feb 14 11:22:37     info.__expected_traceback,
Feb 14 11:22:37   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 656, in execute_once
Feb 14 11:22:37     % (test.__name__, text_repr[0])
Feb 14 11:22:37   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/hypothesis/core.py", line 856, in __flaky
Feb 14 11:22:37     raise Flaky(message)
Feb 14 11:22:37 hypothesis.errors.Flaky: Hypothesis test_avg_pool3d_nhwc(self=<__main__.TestQuantizedOps testMethod=test_avg_pool3d_nhwc>, X=(array([[[[[0., 0., 0., 0., 0.],
```

**Comment by lly-zero-one** [comment]
Currently, the channel last API does not work for the 3D case. I am still waiting for the landing of fix.

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp (hunk 3 above)
Comment (lly-zero-one) [reply aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp]: For 2d, we should have B: batch, C: channel. I think D is for 3-d case.

**Review by jamesr66a — APPROVED** [review]
Seems fine to me. Did you do any perf benchmarking?

**Comment by lly-zero-one** [comment]
> Seems fine to me. Did you do any perf benchmarking?

Yes, will post the result in this PR.

**Review by jianyuh — APPROVED** [review]
LGTM!

**Pull request closed without merging**